File: tauri/src/midi/pedals/reverse_mode_c/mapper.rs
// Reverse Mode C MIDI CC mapping - infrastructure layer

use super::types::{
    ReverseModeCParameter, ReverseModeCState,
    sequence_spacing_to_cc,
};
use std::collections::HashMap;

pub const CC_PRESET_SAVE: u8 = 111;

impl ReverseModeCParameter {
    /// Get the CC number for this parameter
    pub fn cc_number(&self) -> u8 {
        match self {
            // Main knobs
            ReverseModeCParameter::Time(_) => 14,
            ReverseModeCParameter::Mix(_) => 15,
            ReverseModeCParameter::Feedback(_) => 16,
            ReverseModeCParameter::Offset(_) => 17,
            ReverseModeCParameter::Balance(_) => 18,
            ReverseModeCParameter::Filter(_) => 19,
            ReverseModeCParameter::RampSpeed(_) => 20,

            // Three-position toggles
            ReverseModeCParameter::ModSync(_) => 21,
            ReverseModeCParameter::ModType(_) => 22,
            ReverseModeCParameter::SequenceMode(_) => 23,

            // Hidden options
            ReverseModeCParameter::SequencerSubdivision(_) => 24,
            ReverseModeCParameter::RampingWaveform(_) => 25,
            ReverseModeCParameter::ModDepth(_) => 27,
            ReverseModeCParameter::ModRate(_) => 28,
            ReverseModeCParameter::OctaveType(_) => 31,
            ReverseModeCParameter::SequenceSpacing(_) => 33,

            // Footswitches
            ReverseModeCParameter::Bypass(_) => 102,
            ReverseModeCParameter::Tap(_) => 103,
            ReverseModeCParameter::AltMode(_) => 104,
            ReverseModeCParameter::Freeze(_) => 105,
            ReverseModeCParameter::HalfSpeed(_) => 106,

            // DIP switches - Left bank
            ReverseModeCParameter::DipTime(_) => 61,
            ReverseModeCParameter::DipOffset(_) => 62,
            ReverseModeCParameter::DipBalance(_) => 63,
            ReverseModeCParameter::DipFilter(_) => 64,
            ReverseModeCParameter::DipFeed(_) => 65,
            ReverseModeCParameter::DipBounce(_) => 66,
            ReverseModeCParameter::DipSweep(_) => 67,
            ReverseModeCParameter::DipPolarity(_) => 68,

            // DIP switches - Right bank
            ReverseModeCParameter::DipSwap(_) => 71,
            ReverseModeCParameter::DipMiso(_) => 72,
            ReverseModeCParameter::DipSpread(_) => 73,
            ReverseModeCParameter::DipTrails(_) => 74,
            ReverseModeCParameter::DipLatch(_) => 75,
            ReverseModeCParameter::DipFeedType(_) => 76,
            ReverseModeCParameter::DipFadeType(_) => 77,
            ReverseModeCParameter::DipModType(_) => 78,

            // Utility
            ReverseModeCParameter::MidiClockIgnore(_) => 51,
            ReverseModeCParameter::RampBounce(_) => 52,
            ReverseModeCParameter::DryKill(_) => 57,
            ReverseModeCParameter::Expression(_) => 100,
            ReverseModeCParameter::PresetSave(_) => 111,
        }
    }

    /// Get the CC value for this parameter
    pub fn cc_value(&self) -> u8 {
        match self {
            // Continuous parameters
            ReverseModeCParameter::Time(v)
            | ReverseModeCParameter::Mix(v)
            | ReverseModeCParameter::Feedback(v)
            | ReverseModeCParameter::Offset(v)
            | ReverseModeCParameter::Balance(v)
            | ReverseModeCParameter::Filter(v)
            | ReverseModeCParameter::RampSpeed(v)
            | ReverseModeCParameter::SequencerSubdivision(v)
            | ReverseModeCParameter::RampingWaveform(v)
            | ReverseModeCParameter::ModDepth(v)
            | ReverseModeCParameter::ModRate(v)
            | ReverseModeCParameter::Expression(v)
            | ReverseModeCParameter::PresetSave(v) => *v,

            // Enum parameters
            ReverseModeCParameter::ModSync(m) => m.to_cc_value(),
            ReverseModeCParameter::ModType(m) => m.to_cc_value(),
            ReverseModeCParameter::SequenceMode(m) => m.to_cc_value(),
            ReverseModeCParameter::OctaveType(o) => o.to_cc_value(),
            ReverseModeCParameter::DipSweep(d) => d.to_cc_value(),
            ReverseModeCParameter::DipPolarity(p) => p.to_cc_value(),

            // sequence_spacing: false=Rest(1), true=Skip(2)
            ReverseModeCParameter::SequenceSpacing(active) => sequence_spacing_to_cc(*active),

            // Binary parameters
            ReverseModeCParameter::Bypass(b)
            | ReverseModeCParameter::Tap(b)
            | ReverseModeCParameter::AltMode(b)
            | ReverseModeCParameter::Freeze(b)
            | ReverseModeCParameter::HalfSpeed(b)
            | ReverseModeCParameter::DipTime(b)
            | ReverseModeCParameter::DipOffset(b)
            | ReverseModeCParameter::DipBalance(b)
            | ReverseModeCParameter::DipFilter(b)
            | ReverseModeCParameter::DipFeed(b)
            | ReverseModeCParameter::DipBounce(b)
            | ReverseModeCParameter::DipSwap(b)
            | ReverseModeCParameter::DipMiso(b)
            | ReverseModeCParameter::DipSpread(b)
            | ReverseModeCParameter::DipTrails(b)
            | ReverseModeCParameter::DipLatch(b)
            | ReverseModeCParameter::DipFeedType(b)
            | ReverseModeCParameter::DipFadeType(b)
            | ReverseModeCParameter::DipModType(b)
            | ReverseModeCParameter::MidiClockIgnore(b)
            | ReverseModeCParameter::RampBounce(b)
            | ReverseModeCParameter::DryKill(b) => {
                if *b { 127 } else { 0 }
            }
        }
    }

    /// Get a human-readable name for this parameter
    pub fn name(&self) -> &'static str {
        match self {
            ReverseModeCParameter::Time(_) => "Time",
            ReverseModeCParameter::Mix(_) => "Mix",
            ReverseModeCParameter::Feedback(_) => "Feedback",
            ReverseModeCParameter::Offset(_) => "Offset",
            ReverseModeCParameter::Balance(_) => "Balance",
            ReverseModeCParameter::Filter(_) => "Filter",
            ReverseModeCParameter::RampSpeed(_) => "Ramp Speed",
            ReverseModeCParameter::ModSync(_) => "Mod Sync",
            ReverseModeCParameter::ModType(_) => "Mod Type",
            ReverseModeCParameter::SequenceMode(_) => "Sequence Mode",
            ReverseModeCParameter::SequencerSubdivision(_) => "Sequencer Subdivision",
            ReverseModeCParameter::RampingWaveform(_) => "Ramping Waveform",
            ReverseModeCParameter::ModDepth(_) => "Mod Depth",
            ReverseModeCParameter::ModRate(_) => "Mod Rate",
            ReverseModeCParameter::OctaveType(_) => "Octave Type",
            ReverseModeCParameter::SequenceSpacing(_) => "Sequence Spacing",
            ReverseModeCParameter::Bypass(_) => "Bypass",
            ReverseModeCParameter::Tap(_) => "Tap",
            ReverseModeCParameter::AltMode(_) => "Alt Mode",
            ReverseModeCParameter::Freeze(_) => "Freeze",
            ReverseModeCParameter::HalfSpeed(_) => "Half Speed",
            ReverseModeCParameter::DipTime(_) => "DIP: Time",
            ReverseModeCParameter::DipOffset(_) => "DIP: Offset",
            ReverseModeCParameter::DipBalance(_) => "DIP: Balance",
            ReverseModeCParameter::DipFilter(_) => "DIP: Filter",
            ReverseModeCParameter::DipFeed(_) => "DIP: Feed",
            ReverseModeCParameter::DipBounce(_) => "DIP: Bounce",
            ReverseModeCParameter::DipSweep(_) => "DIP: Sweep",
            ReverseModeCParameter::DipPolarity(_) => "DIP: Polarity",
            ReverseModeCParameter::DipSwap(_) => "DIP: Swap",
            ReverseModeCParameter::DipMiso(_) => "DIP: Miso",
            ReverseModeCParameter::DipSpread(_) => "DIP: Spread",
            ReverseModeCParameter::DipTrails(_) => "DIP: Trails",
            ReverseModeCParameter::DipLatch(_) => "DIP: Latch",
            ReverseModeCParameter::DipFeedType(_) => "DIP: Feed Type",
            ReverseModeCParameter::DipFadeType(_) => "DIP: Fade Type",
            ReverseModeCParameter::DipModType(_) => "DIP: Mod Type",
            ReverseModeCParameter::MidiClockIgnore(_) => "MIDI Clock Ignore",
            ReverseModeCParameter::RampBounce(_) => "Ramp Bounce",
            ReverseModeCParameter::DryKill(_) => "Dry Kill",
            ReverseModeCParameter::Expression(_) => "Expression",
            ReverseModeCParameter::PresetSave(_) => "Preset Save",
        }
    }
}

impl ReverseModeCState {
    /// Convert the current state to a map of CC numbers -> CC values.
    /// Used when recalling a full preset (sending all parameters at once).
    pub fn to_cc_map(&self) -> HashMap<u8, u8> {
        let mut map = HashMap::new();

        // Main knobs
        map.insert(14, self.time);
        map.insert(15, self.mix);
        map.insert(16, self.feedback);
        map.insert(17, self.offset);
        map.insert(18, self.balance);
        map.insert(19, self.filter);
        map.insert(20, self.ramp_speed);

        // Three-position toggles
        map.insert(21, self.mod_sync.to_cc_value());
        map.insert(22, self.mod_type.to_cc_value());
        map.insert(23, self.sequence_mode.to_cc_value());

        // Hidden options
        map.insert(24, self.sequencer_subdivision);
        map.insert(25, self.ramping_waveform);
        map.insert(27, self.mod_depth);
        map.insert(28, self.mod_rate);
        map.insert(31, self.octave_type.to_cc_value());
        map.insert(33, sequence_spacing_to_cc(self.sequence_spacing));

        // Footswitches
        map.insert(102, if self.bypass { 127 } else { 0 });
        map.insert(103, if self.tap { 127 } else { 0 });
        map.insert(104, if self.alt_mode { 127 } else { 0 });
        map.insert(105, if self.freeze { 127 } else { 0 });
        map.insert(106, if self.half_speed { 127 } else { 0 });

        // DIP switches - Left bank
        map.insert(61, if self.dip_time { 127 } else { 0 });
        map.insert(62, if self.dip_offset { 127 } else { 0 });
        map.insert(63, if self.dip_balance { 127 } else { 0 });
        map.insert(64, if self.dip_filter { 127 } else { 0 });
        map.insert(65, if self.dip_feed { 127 } else { 0 });
        map.insert(66, if self.dip_bounce { 127 } else { 0 });
        map.insert(67, self.dip_sweep.to_cc_value());
        map.insert(68, self.dip_polarity.to_cc_value());

        // DIP switches - Right bank
        map.insert(71, if self.dip_swap { 127 } else { 0 });
        map.insert(72, if self.dip_miso { 127 } else { 0 });
        map.insert(73, if self.dip_spread { 127 } else { 0 });
        map.insert(74, if self.dip_trails { 127 } else { 0 });
        map.insert(75, if self.dip_latch { 127 } else { 0 });
        map.insert(76, if self.dip_feed_type { 127 } else { 0 });
        map.insert(77, if self.dip_fade_type { 127 } else { 0 });
        map.insert(78, if self.dip_mod_type { 127 } else { 0 });

        // Utility
        map.insert(51, if self.midi_clock_ignore { 127 } else { 0 });
        map.insert(52, if self.ramp_bounce { 127 } else { 0 });
        map.insert(57, if self.dry_kill { 127 } else { 0 });
        map.insert(100, self.expression);

        map
    }
}

File: apps/web/src/lib/midi/pedals/reverse-mode-c/mapper.ts
// MIDI CC mapping for Reverse Mode C parameters
// Ported from apps/desktop/tauri/src/midi/pedals/reverse_mode_c/mapper.rs

import type {
  ReverseModeCParameter,
  ReverseModeCState,
  ModSync,
  ModType,
  SequenceMode,
  OctaveType,
  SweepDirection,
  Polarity,
} from './types';

function modSyncToCCValue(m: ModSync): number {
  switch (m) { case 'Sync': return 1; case 'Off': return 2; case 'Free': return 3; }
}
function modSyncFromCCValue(v: number): ModSync {
  if (v === 2) return 'Off'; if (v === 3) return 'Free'; return 'Sync';
}

function modTypeToCCValue(m: ModType): number {
  switch (m) { case 'Vib': return 1; case 'Trem': return 2; case 'Freq': return 3; }
}
function modTypeFromCCValue(v: number): ModType {
  if (v === 2) return 'Trem'; if (v === 3) return 'Freq'; return 'Vib';
}

function sequenceModeToCCValue(m: SequenceMode): number {
  switch (m) { case 'Run': return 1; case 'Off': return 2; case 'Env': return 3; }
}
function sequenceModeFromCCValue(v: number): SequenceMode {
  if (v === 2) return 'Off'; if (v === 3) return 'Env'; return 'Run';
}

function octaveTypeToCCValue(o: OctaveType): number {
  switch (o) { case 'OctDown': return 1; case 'BothOct': return 2; case 'OctUp': return 3; }
}
function octaveTypeFromCCValue(v: number): OctaveType {
  if (v === 2) return 'BothOct'; if (v === 3) return 'OctUp'; return 'OctDown';
}

// CC 33: sequence_spacing (false=Rest → 1, true=Skip → 2)
function sequenceSpacingToCCValue(active: boolean): number { return active ? 2 : 1; }
function sequenceSpacingFromCCValue(v: number): boolean { return v >= 2; }

function sweepDirectionToCCValue(d: SweepDirection): number { return d === 'Bottom' ? 0 : 127; }
function sweepDirectionFromCCValue(v: number): SweepDirection { return v < 64 ? 'Bottom' : 'Top'; }
function polarityToCCValue(p: Polarity): number { return p === 'Forward' ? 0 : 127; }
function polarityFromCCValue(v: number): Polarity { return v < 64 ? 'Forward' : 'Reverse'; }

export function parameterCCNumber(param: ReverseModeCParameter): number {
  if ('Time' in param)                 return 14;
  if ('Mix' in param)                  return 15;
  if ('Feedback' in param)             return 16;
  if ('Offset' in param)               return 17;
  if ('Balance' in param)              return 18;
  if ('Filter' in param)               return 19;
  if ('RampSpeed' in param)            return 20;
  if ('ModSync' in param)              return 21;
  if ('ModType' in param)              return 22;
  if ('SequenceMode' in param)         return 23;
  if ('SequencerSubdivision' in param) return 24;
  if ('RampingWaveform' in param)      return 25;
  if ('ModDepth' in param)             return 27;
  if ('ModRate' in param)              return 28;
  if ('OctaveType' in param)           return 31;
  if ('SequenceSpacing' in param)      return 33;
  if ('MidiClockIgnore' in param)      return 51;
  if ('RampBounce' in param)           return 52;
  if ('DryKill' in param)              return 57;
  if ('DipTime' in param)              return 61;
  if ('DipOffset' in param)            return 62;
  if ('DipBalance' in param)           return 63;
  if ('DipFilter' in param)            return 64;
  if ('DipFeed' in param)              return 65;
  if ('DipBounce' in param)            return 66;
  if ('DipSweep' in param)             return 67;
  if ('DipPolarity' in param)          return 68;
  if ('DipSwap' in param)              return 71;
  if ('DipMiso' in param)              return 72;
  if ('DipSpread' in param)            return 73;
  if ('DipTrails' in param)            return 74;
  if ('DipLatch' in param)             return 75;
  if ('DipFeedType' in param)          return 76;
  if ('DipFadeType' in param)          return 77;
  if ('DipModType' in param)           return 78;
  if ('Expression' in param)           return 100;
  if ('Bypass' in param)               return 102;
  if ('Tap' in param)                  return 103;
  if ('AltMode' in param)              return 104;
  if ('Freeze' in param)               return 105;
  if ('HalfSpeed' in param)            return 106;
  if ('PresetSave' in param)           return 111;
  throw new Error(`Unknown ReverseModeCParameter: ${JSON.stringify(param)}`);
}

export function parameterCCValue(param: ReverseModeCParameter): number {
  if ('Time' in param)                 return param.Time;
  if ('Mix' in param)                  return param.Mix;
  if ('Feedback' in param)             return param.Feedback;
  if ('Offset' in param)               return param.Offset;
  if ('Balance' in param)              return param.Balance;
  if ('Filter' in param)               return param.Filter;
  if ('RampSpeed' in param)            return param.RampSpeed;
  if ('SequencerSubdivision' in param) return param.SequencerSubdivision;
  if ('RampingWaveform' in param)      return param.RampingWaveform;
  if ('ModDepth' in param)             return param.ModDepth;
  if ('ModRate' in param)              return param.ModRate;
  if ('Expression' in param)           return param.Expression;
  if ('PresetSave' in param)           return param.PresetSave;
  if ('ModSync' in param)              return modSyncToCCValue(param.ModSync);
  if ('ModType' in param)              return modTypeToCCValue(param.ModType);
  if ('SequenceMode' in param)         return sequenceModeToCCValue(param.SequenceMode);
  if ('OctaveType' in param)           return octaveTypeToCCValue(param.OctaveType);
  if ('SequenceSpacing' in param)      return sequenceSpacingToCCValue(param.SequenceSpacing);
  if ('DipSweep' in param)             return sweepDirectionToCCValue(param.DipSweep);
  if ('DipPolarity' in param)          return polarityToCCValue(param.DipPolarity);
  if ('Bypass' in param)               return param.Bypass ? 127 : 0;
  if ('Tap' in param)                  return param.Tap ? 127 : 0;
  if ('AltMode' in param)              return param.AltMode ? 127 : 0;
  if ('Freeze' in param)               return param.Freeze ? 127 : 0;
  if ('HalfSpeed' in param)            return param.HalfSpeed ? 127 : 0;
  if ('DipTime' in param)              return param.DipTime ? 127 : 0;
  if ('DipOffset' in param)            return param.DipOffset ? 127 : 0;
  if ('DipBalance' in param)           return param.DipBalance ? 127 : 0;
  if ('DipFilter' in param)            return param.DipFilter ? 127 : 0;
  if ('DipFeed' in param)              return param.DipFeed ? 127 : 0;
  if ('DipBounce' in param)            return param.DipBounce ? 127 : 0;
  if ('DipSwap' in param)              return param.DipSwap ? 127 : 0;
  if ('DipMiso' in param)              return param.DipMiso ? 127 : 0;
  if ('DipSpread' in param)            return param.DipSpread ? 127 : 0;
  if ('DipTrails' in param)            return param.DipTrails ? 127 : 0;
  if ('DipLatch' in param)             return param.DipLatch ? 127 : 0;
  if ('DipFeedType' in param)          return param.DipFeedType ? 127 : 0;
  if ('DipFadeType' in param)          return param.DipFadeType ? 127 : 0;
  if ('DipModType' in param)           return param.DipModType ? 127 : 0;
  if ('MidiClockIgnore' in param)      return param.MidiClockIgnore ? 127 : 0;
  if ('RampBounce' in param)           return param.RampBounce ? 127 : 0;
  if ('DryKill' in param)              return param.DryKill ? 127 : 0;
  throw new Error(`Unknown ReverseModeCParameter: ${JSON.stringify(param)}`);
}

export function stateToCCMap(state: ReverseModeCState): Map<number, number> {
  const map = new Map<number, number>();
  map.set(14, state.time);
  map.set(15, state.mix);
  map.set(16, state.feedback);
  map.set(17, state.offset);
  map.set(18, state.balance);
  map.set(19, state.filter);
  map.set(20, state.ramp_speed);
  map.set(21, modSyncToCCValue(state.mod_sync));
  map.set(22, modTypeToCCValue(state.mod_type));
  map.set(23, sequenceModeToCCValue(state.sequence_mode));
  map.set(24, state.sequencer_subdivision);
  map.set(25, state.ramping_waveform);
  map.set(27, state.mod_depth);
  map.set(28, state.mod_rate);
  map.set(31, octaveTypeToCCValue(state.octave_type));
  map.set(33, sequenceSpacingToCCValue(state.sequence_spacing));
  map.set(51, state.midi_clock_ignore ? 127 : 0);
  map.set(52, state.ramp_bounce ? 127 : 0);
  map.set(57, state.dry_kill ? 127 : 0);
  map.set(61, state.dip_time ? 127 : 0);
  map.set(62, state.dip_offset ? 127 : 0);
  map.set(63, state.dip_balance ? 127 : 0);
  map.set(64, state.dip_filter ? 127 : 0);
  map.set(65, state.dip_feed ? 127 : 0);
  map.set(66, state.dip_bounce ? 127 : 0);
  map.set(67, sweepDirectionToCCValue(state.dip_sweep));
  map.set(68, polarityToCCValue(state.dip_polarity));
  map.set(71, state.dip_swap ? 127 : 0);
  map.set(72, state.dip_miso ? 127 : 0);
  map.set(73, state.dip_spread ? 127 : 0);
  map.set(74, state.dip_trails ? 127 : 0);
  map.set(75, state.dip_latch ? 127 : 0);
  map.set(76, state.dip_feed_type ? 127 : 0);
  map.set(77, state.dip_fade_type ? 127 : 0);
  map.set(78, state.dip_mod_type ? 127 : 0);
  map.set(100, state.expression);
  map.set(102, state.bypass ? 127 : 0);
  map.set(103, state.tap ? 127 : 0);
  map.set(104, state.alt_mode ? 127 : 0);
  map.set(105, state.freeze ? 127 : 0);
  map.set(106, state.half_speed ? 127 : 0);
  return map;
}

export function ccToParameter(cc: number, value: number): ReverseModeCParameter | null {
  switch (cc) {
    case 14:  return { Time: value };
    case 15:  return { Mix: value };
    case 16:  return { Feedback: value };
    case 17:  return { Offset: value };
    case 18:  return { Balance: value };
    case 19:  return { Filter: value };
    case 20:  return { RampSpeed: value };
    case 21:  return { ModSync: modSyncFromCCValue(value) };
    case 22:  return { ModType: modTypeFromCCValue(value) };
    case 23:  return { SequenceMode: sequenceModeFromCCValue(value) };
    case 24:  return { SequencerSubdivision: value };
    case 25:  return { RampingWaveform: value };
    case 27:  return { ModDepth: value };
    case 28:  return { ModRate: value };
    case 31:  return { OctaveType: octaveTypeFromCCValue(value) };
    case 33:  return { SequenceSpacing: sequenceSpacingFromCCValue(value) };
    case 51:  return { MidiClockIgnore: value >= 64 };
    case 52:  return { RampBounce: value >= 64 };
    case 57:  return { DryKill: value >= 64 };
    case 61:  return { DipTime: value >= 64 };
    case 62:  return { DipOffset: value >= 64 };
    case 63:  return { DipBalance: value >= 64 };
    case 64:  return { DipFilter: value >= 64 };
    case 65:  return { DipFeed: value >= 64 };
    case 66:  return { DipBounce: value >= 64 };
    case 67:  return { DipSweep: sweepDirectionFromCCValue(value) };
    case 68:  return { DipPolarity: polarityFromCCValue(value) };
    case 71:  return { DipSwap: value >= 64 };
    case 72:  return { DipMiso: value >= 64 };
    case 73:  return { DipSpread: value >= 64 };
    case 74:  return { DipTrails: value >= 64 };
    case 75:  return { DipLatch: value >= 64 };
    case 76:  return { DipFeedType: value >= 64 };
    case 77:  return { DipFadeType: value >= 64 };
    case 78:  return { DipModType: value >= 64 };
    case 100: return { Expression: value };
    case 102: return { Bypass: value >= 64 };
    case 103: return { Tap: value >= 64 };
    case 104: return { AltMode: value >= 64 };
    case 105: return { Freeze: value >= 64 };
    case 106: return { HalfSpeed: value >= 64 };
    case 111: return { PresetSave: value };
    default:  return null;
  }
}
